README.md
# F1 Delta Time, Core Metadata library

## Core metadata

### Core attributes

An F1 Delta Time NFT comes with a number of core attributes embedded within the value of the token identifier. These attributes comprise what we call the token's _Core Metadata_. There are two main benefits for using this structure:
1. the core attributes are set at creation and forever immutable,
2. the tokens can be interpreted on-chain. For example, a contract could provide a service in exchange for a token but only if this token is a _Car_ with a rarity between 7 and 9.

Here is the current list of the core attributes:

| Attribute Name | Binary Position | Data Type | Required | Season Specific | Collection Identifier | Description |
| :---           | :---            | :---      | :---:    | :---:           | :---                  | :---        |
| nonFungibleFlag  | 255  | bit    | no  | no  | yes | Used to distinguish fungible and non-fungible tokens. A value of _1_ represents a non-fungible token. |
| padding1         | 248  | uint7  | no  | N/A | yes | Reserved for later use. |
| typeId           | 240  | uint8  | yes | no  | yes | Numerical representation of the type of the token, for example _Car_, _Driver_ or _Gears_. |
| subTypeId        | 232  | uint8  | no  | yes | yes | Numerical representation of the subType of the token, when applicable. For example _Gears_ have different possible subTypes such as _Helmet_ or _Gloves_. |
| seasonId         | 224  | uint8  | yes | no  | yes | Numerical representation of the season of the token, for example _2019_. |
| padding2         | 212  | uint32 | no  | N/A | no  | Reserved for later use. |
| trackId          | 200  | uint8  | no  | yes | no  | Numerical representation of the track of the token, for example _Circuit de Monaco_. |
| modelId          | 192  | uint8  | no  | yes | no  | Numerical representation of the model of the token, for example _Satsuma_. |
| teamId           | 184  | uint8  | no  | yes | no  | Numerical representation of the team of the token, for example _Alfa Romeo Racing_. |
| rarity           | 176  | uint8  | yes | no  | no  | The rarity level of the token. |
| countryId          | 168  | uint8  | no  | yes | no  | Numerical representation of the country of the token, for example _Monaco_. |
| labelId          | 152  | uint16 | no  | no  | no  | Numerical representation of label of the token. Label are used for specific tokens and is intended to be used for collection and gamfication. |
| driverId     | 136  | uint16 | no  | no  | no  | The driver number for a _Driver_. |
| stat1            | 120  | uint16 | no  | yes | no  | The _Top Speed_ of a _Car_/_Part_/_Tyre_ **or** the _Stamina_ of a _Driver_/_Gear_. |
| stat2            | 104  | uint16 | no  | yes | no  | The _Acceleration_ of a _Car_/_Part_/_Tyre_ **or** the _Aggression_ of a _Driver_/_Gear_. |
| stat3            | 88   | uint16 | no  | yes | no  | The _Grip_ of a _Car_/_Part_/_Tyre_ **or** the _Concentration_ of a _Driver_/_Gear_. |
| luck             | 72   | uint16 | no  | yes | no  | The luck of the token. |
| effect           | 64   | uint8  | no  | yes | no  | The effect of the token. |
| special1         | 56   | uint8  | no  | yes | no  | The first special of the token. |
| special2         | 48   | uint8  | no  | yes | no  | The second special of the token. |
| counter          | 0    | uint48 | no  | yes | no  | Field used to ensure unicity of the tokens in case 2 tokens would have exactly the same other attributes. |

`Required` indicates whether the attribute must have a value different from zero.

`Season Specific` indicates whether the attribute's interpretation can vary between seasons.

`Collection Identifier` indicates whether the attribute is part of the collection identifier. Collection identifiers are primarily composed of the _typeId_, _subTypeId_ and _seasonId_. For example _Car 2019_ or _Intermediate Tyres 2020_.

Some attributes may or may not be needed, depending on the type, subType and season of the token (refer to the season-specific mapping files for more details).

### Mapped core attributes

Core attributes are encoded as numerical values. Those which name ends with `Id` are _unmapped_ and can be mapped to their text value using the mappings contained in this library. For example, typeId 1 can be mapped to the text value _Car_.

## Library usage

### `module.metadata` object

A set of functions which help with the manipulation of token identifiers and metadata objects.

#### `coreMetadataFromId(id)` and `fullMetadataFromId(id, [network])` functions

Build metadata objects from token identifiers.

```javascript
const { coreMetadataFromId, fullMetadataFromId } = require('@animoca/f1dt-core_metadata').metadata;

const tokenId = '57897811519642769433138067471762254623735906850517137802921006713614358282351'; // the 1-1-1
const coreMetadata = coreMetadataFromId(tokenId);
console.log(coreMetadata);
const fullMetadata = fullMetadataFromId(tokenId);
console.log(fullMetadata);
```

#### `idFromCoreMetadata(coreMetadata)` function

Build token identifiers from core attributes (mapped and/or unmapped).

```javascript
const { idFromCoreMetadata } = require('@animoca/f1dt-core_metadata').metadata;
const tokenId1 = idFromCoreMetadata({
    typeId: '1', // Car
    subTypeId: '0',
    seasonId: '2', // 2019
});
const tokenId2 = idFromCoreMetadata({
    typeId: '1', // Car
    subTypeId: '0',
    season: '2019',
});
const tokenId3 = idFromCoreMetadata({
    type: 'Car',
    subTypeId: 'None',
    seasonId: '2019',
});
console.log(tokenId1, tokenId2, tokenId3);
// tokenId1 == tokenId2 == tokenId3
```

### `module.constants` object

Some project constants such as the bits layout object and the fixed number of bits in the collection identifiers.

### `module.mappings` object

All the mappings necessary to manipulate metadata divided in common mappings and season-specific mappings.

Retrieve the name of a team from its id for season 2019:
```javascript
const seasonMappings = require('@animoca/f1dt-core_metadata').mappings['2019'];

```

Retrieve the Tier for a rarity:
```javascript

```

### `module.collections` object

The lists of existing collections.

### Scripts

#### Generate metadata from a token identifier
```bash
node scripts/metadataFromId.js -i <id> -n <network> [--full]
```


CHANGELOG.md
# Changelog

## 2.2.1
* Updated Australia track segment name and track name
* Updated Italy track segment name and description

## 2.2.0
* Added trackSegment.eventSeason core metadata.
* Track segement names are based on eventSeason.

## 2.1.6
* Added missing description for Bahrain and Australia track segments.

## 2.1.5
* Added Bahrain and Australia track segments.

## 2.1.4
* Fixed track nft types in opensea attribute are messed up.

## 2.1.3
* Updated track nft type name on opensea from 'track' to 'event segment'

## 2.1.2
* Updated descriptions for Silverstone, Monza and Shanghai tracks

## 2.1.1
* Updated image name for all existing tracks

## 2.1.0
* Updated Track name for Silverstone
* Added Shanghai track
* Added/Updated track segment name for Silverstone, Monza and Shanghai

## 2.0.9
* Minor fix on Track.js countryId should use the countryId in Countries.js instead of an object

## 2.0.8
* Fixed China Azure Dragon edition youtube url is using a shortern youtube url

## 2.0.7
* Fixed FirstEdition car images should be align with the team car image

## 2.0.6
* Fixed First edition Cars are missing youtube_url attribute

## 2.0.5
* Revert tokenId: 57899578366707547814966713827535462487986658084231150927309255110625459205623 from team Aston Martin back to Toro Rosso

## 2.0.4
* Fixed Lewis Hamilton description

## 2.0.3
* Updated 2020 Team Drivers description

## 2.0.2
* Added Belgium Track Segment

## 2.0.1
* Updated generic driver description
* Fixed image override settings is not working
* Fixed Kimi Räikkönen and Sergio Pérez image url is not correct

## 2.0.0
* Added 2020 NFTs
* Code cleanup (removing extendedMeta in TokenTypes)

## 1.2.1
* Migrate image url to another CDN
* Forced Apex car to generate image url using getImageUrl

## 1.1.8
* Updated youtube url for Australia Edition 2020
* Added 70th Anniversary Edition Apex car

## 1.1.7
* Updated Track segment name and description

## 1.1.6
### Improvements
* Removed opensea attributes `earning_share` for Tracks 

## 1.1.5

### Bugfix
* Fix the ES6 runtime issue in TrackSegment.js

## 1.1.0

### New Features
* Add Track Segment layout for track NFTs
* Add Bahrain Edition Apex Car NFT
* Add Monaco Track Segments NFT

### Bugfix
* Fix a crash during validation of the Core Metadata
* Fix the original trackId using is actually countryId
* Should skip validating season metadata when `season` field is 0

### Improvements
* House keeping on the folder structure and some code cleanup

## 1.0.4

### Bugfix
* Fixed a bug showing a wrong rarity for pre-crates tokens.
* Swap two driver's team and add new entries in the repair list.

### Improvements
 * Add `collection` field in full metadata.

## 1.0.3

### Bugfix
 * Fix the collections flattening.

## 1.0.2

### New features
 * Optional validation step for token id creation.

## 1.0.1

### Breaking changes
 * Major code refactor.

### New features
 * Token id creation from metadata.
 * Metadata validation.

### Improvements
 * Improved documentation.
 * Integrated `prettier`.

### Bugfixes
* Several bugfixes.

## 0.1.2

### Bugfixes
* Fixed image url generation for tyres

### Chores
* Removed `console.log` from `metadata/image.js`

## 0.1.1

### Bugfixes
* Fix require paths for seasons.

## 0.1.0

### New Features
* Added `collections` with support for inventory metadata layouts.
* Added season 2020 mappings (WIP).

### Chores
 * Moved `metadataFromId.js` to `scripts` folder.

## 0.0.2

### Bugfixes
* Fixed wrong require path.

## 0.0.1
* Initial commit.


src/constants.js
const {validateBitsLayout} = require('../src/utils/layout');

const TokenIdLength = 256;
const CollectionMaskLength = 32;

const TokenBitsLayout = [
  //         COLLECTION ID
  {name: 'nfFlag', bits: 1}, // 255
  {name: 'padding1', bits: 7}, // 248
  {name: 'typeId', bits: 8}, // 240
  {name: 'subTypeId', bits: 8}, // 232
  {name: 'seasonId', bits: 8}, // 224
  //         OTHER ATTRIBUTES
  {name: 'padding2', bits: 24}, // 224
  {name: 'modelId', bits: 8}, // 192
  {name: 'teamId', bits: 8}, // 184
  {name: 'rarity', bits: 8}, // 176
  {name: 'countryId', bits: 8}, // 168
  {name: 'labelId', bits: 16}, // 152
  {name: 'driverId', bits: 16}, // 136
  {name: 'racing.stat1', bits: 16}, // 120
  {name: 'racing.stat2', bits: 16}, // 104
  {name: 'racing.stat3', bits: 16}, // 88
  {name: 'racing.luck', bits: 16}, // 72
  {name: 'racing.effect', bits: 8}, // 64
  {name: 'racing.special1', bits: 8}, // 56
  {name: 'racing.special2', bits: 8}, // 48
  {name: 'counter', bits: 48}, // 0
].reverse();
Object.freeze(TokenBitsLayout);
validateBitsLayout(TokenBitsLayout);

const TrackSegmentTokenBitsLayout = [
  //         COLLECTION ID
  {name: 'nfFlag', bits: 1}, // 255
  {name: 'padding1', bits: 7}, // 248
  {name: 'typeId', bits: 8}, // 240
  {name: 'subTypeId', bits: 8}, // 232
  {name: 'seasonId', bits: 8}, // 224

  //         Other Attributes
  {name: 'padding2', bits: 12}, // 212
  {name: 'trackId', bits: 12}, // 200
  {name: 'modelId', bits: 8}, // 192
  {name: 'teamId', bits: 8}, // 184
  {name: 'rarity', bits: 8}, // 176
  {name: 'countryId', bits: 8}, // 168
  {name: 'labelId', bits: 16}, // 152
  {name: 'driverId', bits: 16}, // 136

  //          Track Attributes
  {name: 'trackSegment.zoneId', bits: 8}, // 128
  {name: 'trackSegment.segmentId', bits: 8}, // 120
  {name: 'trackSegment.earnings', bits: 32}, // 88
  {name: 'trackSegment.eventSeason', bits: 16}, // 72
  {name: 'padding3', bits: 24}, // 48
  {name: 'counter', bits: 48}, // 0
].reverse();
Object.freeze(TrackSegmentTokenBitsLayout);
validateBitsLayout(TrackSegmentTokenBitsLayout);

const CollectionEncodingBitsLayout = [
  {name: 'nfFlag', bits: 1}, // 255
  {name: 'padding1', bits: 7}, // 248
  {name: 'typeId', bits: 8}, // 240
  {name: 'subTypeId', bits: 8}, // 232
  {name: 'seasonId', bits: 8}, // 224
  {name: '_', bits: TokenIdLength - CollectionMaskLength}, // 0
].reverse();
Object.freeze(CollectionEncodingBitsLayout);
validateBitsLayout(CollectionEncodingBitsLayout);

const CollectionDecodingBitsLayout = [
  {name: 'collectionId', bits: CollectionMaskLength}, // 224
  {name: '_', bits: TokenIdLength - CollectionMaskLength}, // 0
].reverse();
Object.freeze(CollectionDecodingBitsLayout);
validateBitsLayout(CollectionDecodingBitsLayout);

module.exports = {
  TokenIdLength,
  CollectionMaskLength,
  TokenBitsLayout,
  TrackSegmentTokenBitsLayout,
  CollectionDecodingBitsLayout,
  CollectionEncodingBitsLayout,
};
